Kafka Snappy library fails to load when restarting Quarkus in the same JVM

### Describe the bug

I have a project where the tests run with multiple test profiles, which causes a new Quarkus instance for each group of tests (but I believe it's still the same JVM instance). However, the Snappy library that is being loaded cannot be loaded the second time since there is still a lock on the snappyjava.dll file (I think the OS keeps a lock since it's loaded in as a DLL)

### Expected behavior

Snappy loads successfully each Quarkus startup

### Actual behavior

The first time Quarkus starts up successfully. The second time Quarkus starts up, I get the following exception:

```java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:638)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:722)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.AbstractList$RandomAccessSpliterator.forEachRemaining(AbstractList.java:722)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1006)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:285)
	at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:252)
	at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:605)
	at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:655)
	... 1 more
Caused by: java.io.UncheckedIOException: Unable to extract native library snappyjava.dll to C:\Users\<user>\AppData\Local\Temp\snappyjava.dll
	at io.quarkus.kafka.client.runtime.SnappyRecorder.extractLibraryFile(SnappyRecorder.java:54)
	at io.quarkus.kafka.client.runtime.SnappyRecorder.loadSnappy(SnappyRecorder.java:30)
	at io.quarkus.deployment.steps.KafkaProcessor$loadSnappyIfEnabled852439731.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.KafkaProcessor$loadSnappyIfEnabled852439731.deploy(Unknown Source)
	... 8 more
Caused by: java.io.FileNotFoundException: C:\Users\<user>\AppData\Local\Temp\snappyjava.dll (The process cannot access the file because it is being used by another process)
	at java.base/java.io.FileOutputStream.open0(Native Method)
	at java.base/java.io.FileOutputStream.open(FileOutputStream.java:295)
	at java.base/java.io.FileOutputStream.<init>(FileOutputStream.java:236)
	at java.base/java.io.FileOutputStream.<init>(FileOutputStream.java:185)
	at io.quarkus.kafka.client.runtime.SnappyRecorder.extractLibraryFile(SnappyRecorder.java:46)
	... 11 more
```

### How to Reproduce?

1. Create a Quarkus 3.8.2 Maven project with a dependency to quarkus-kafka-client and `quarkus.kafka.snappy.enabled=true`
2. Create 2 tests, each with `@QuarkusTest` and a different `@TestProfile`
3. Run the tests, for example with mvn clean install

### Output of `uname -a` or `ver`

Microsoft Windows [Version 10.0.22631.3296]

### Output of `java -version`

openjdk 20.0.1 2023-04-18 OpenJDK Runtime Environment (build 20.0.1+9-29) OpenJDK 64-Bit Server VM (build 20.0.1+9-29, mixed mode, sharing)

### Quarkus version or git rev

3.8.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.4 (dfbb324ad4a7c8fb0bf182e6d91b0ae20e3d2dd9)

### Additional information

_No response_

--- a/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaBuildTimeConfig.java
+++ b/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaBuildTimeConfig.java
@@ -24,6 +24,14 @@ public class KafkaBuildTimeConfig {
     public boolean snappyEnabled;
 
     /**
+     * Whether to load the Snappy native library from the shared classloader.
+     * This setting is only used in tests if the tests are using different profiles, which would lead to
+     * unsatisfied link errors when loading Snappy.
+     */
+    @ConfigItem(name = "snappy.load-from-shared-classloader", defaultValue = "false")
+    public boolean snappyLoadFromSharedClassLoader;
+
+    /**
      * Configuration for DevServices. DevServices allows Quarkus to automatically start Kafka in dev and test mode.
      */
     @ConfigItem

--- a/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
+++ b/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
@@ -72,6 +72,7 @@ import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.ExtensionSslNativeSupportBuildItem;
 import io.quarkus.deployment.builditem.FeatureBuildItem;
 import io.quarkus.deployment.builditem.IndexDependencyBuildItem;
+import io.quarkus.deployment.builditem.LaunchModeBuildItem;
 import io.quarkus.deployment.builditem.LogCategoryBuildItem;
 import io.quarkus.deployment.builditem.RunTimeConfigurationDefaultBuildItem;
 import io.quarkus.deployment.builditem.RuntimeConfigSetupCompleteBuildItem;
@@ -304,8 +305,12 @@ public class KafkaProcessor {
 
     @BuildStep(onlyIf = HasSnappy.class)
     @Record(ExecutionTime.RUNTIME_INIT)
-    void loadSnappyIfEnabled(SnappyRecorder recorder, KafkaBuildTimeConfig config) {
-        recorder.loadSnappy();
+    void loadSnappyIfEnabled(LaunchModeBuildItem launch, SnappyRecorder recorder, KafkaBuildTimeConfig config) {
+        boolean loadFromSharedClassLoader = false;
+        if (launch.isTest()) {
+            loadFromSharedClassLoader = config.snappyLoadFromSharedClassLoader;
+        }
+        recorder.loadSnappy(loadFromSharedClassLoader);
     }
 
     @Consume(RuntimeConfigSetupCompleteBuildItem.class)

--- a/extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/SnappyLoader.java
+++ b/extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/SnappyLoader.java
@@ -1,0 +1,18 @@
+package io.quarkus.kafka.client.runtime;
+
+import java.io.File;
+
+public class SnappyLoader {
+
+    /*
+     * This class is intended to be loaded from a shared classloader (e.g., the system classloader) to avoid
+     * unsatisfied link errors when the native library is loaded from a different classloader.
+     * See https://github.com/quarkusio/quarkus/issues/39767.
+     *
+     * This class is only used in tests if the `quarkus.kafka.snappy.load-from-shared-classloader=true` is set.
+     */
+    static {
+        File out = SnappyRecorder.getLibraryFile();
+        System.load(out.getAbsolutePath());
+    }
+}

--- a/extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/SnappyRecorder.java
+++ b/extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/SnappyRecorder.java
@@ -8,14 +8,32 @@ import java.io.UncheckedIOException;
 import java.net.URL;
 
 import org.xerial.snappy.OSInfo;
-import org.xerial.snappy.SnappyLoader;
 
+import io.quarkus.runtime.Application;
 import io.quarkus.runtime.annotations.Recorder;
 
 @Recorder
 public class SnappyRecorder {
 
-    public void loadSnappy() {
+    public void loadSnappy(boolean loadFromSharedClassLoader) {
+        if (loadFromSharedClassLoader) {
+            try {
+                Application.class.getClassLoader().loadClass(SnappyLoader.class.getName());
+            } catch (ClassNotFoundException e) {
+                throw new RuntimeException(e);
+            }
+        } else {
+            File out = getLibraryFile();
+            try {
+                System.load(out.getAbsolutePath());
+            } catch (UnsatisfiedLinkError e) {
+                // Try to load the library from the system library path
+                throw new RuntimeException("Failed to load Snappy native library", e);
+            }
+        }
+    }
+
+    static File getLibraryFile() {
         // Resolve the library file name with a suffix (e.g., dll, .so, etc.)
         String snappyNativeLibraryName = System.mapLibraryName("snappyjava");
         String snappyNativeLibraryPath = "/org/xerial/snappy/native/" + OSInfo.getNativeLibFolderPathForCurrentOS();
@@ -27,18 +45,16 @@ public class SnappyRecorder {
             throw new RuntimeException(errorMessage);
         }
 
-        File out = extractLibraryFile(
+        return extractLibraryFile(
                 SnappyLoader.class.getResource(snappyNativeLibraryPath + "/" + snappyNativeLibraryName),
                 snappyNativeLibraryName);
-
-        System.load(out.getAbsolutePath());
     }
 
-    private static boolean hasResource(String path) {
+    static boolean hasResource(String path) {
         return SnappyLoader.class.getResource(path) != null;
     }
 
-    private static File extractLibraryFile(URL library, String name) {
+    static File extractLibraryFile(URL library, String name) {
         String tmp = System.getProperty("java.io.tmpdir");
         File extractedLibFile = new File(tmp, name);
 

